### ISSUE TITLE
30 Minute Compilation Time

### ISSUE THREAD

**Issue by MikeynJerry** [pytorch/xla#5058]
## ❓ Questions and Help

I'm currently trying to train Stable Diffusion v1.5 on a small TPU slice (v4-8) using `diffusers`. Since the beginning, the compile time has taken about 30 minutes across 3 compiles. Originally, the first compile (and iteration) took ~4 minutes, the second ~14 minutes, and the third ~16 minutes. I found an instance of `aten::upsample_nearest2d_backward` in the logs, thinking that was causing my issues. I copied diffusers locally and changed the upsample code so that the results stayed the same and the `aten::...` went away, which caused the first compile to speed up significantly (4 minutes -> 4 seconds), but the second and third compiles are still super slow (~29 minutes combined). There are no further `aten::...` ops that the metrics reporter mentions so I'm unsure where to look. Below I've attached three things:
- [sd.py](https://github.com/pytorch/xla/files/11549356/sd.py.txt): a cleaned up copy of the training script with a bunch of fluff removed (e.g. model loading, etc)
- [metrics.0.txt](https://github.com/pytorch/xla/files/11549452/metrics.0.txt): the metrics reported by the profiler
- [ir.0.txt](https://github.com/pytorch/xla/files/11549453/ir.0.txt): The graph from the three compilations performed on TPU 0 (of 4)

Any ideas or help would be much appreciated! Thank you in advance!

**Comment by JackCaoG** [comment]
Thanks, can you share the dump of HLO instead of IR? You can do that by setting `XLA_SAVE_TENSORS_FMT=hlo` following https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#environment-variables . 

For aten thing, I think what happened is that because of `aten::upsample_nearest2d_backward` we fallback to CPU and cut a single graph into 2 smaller graph, which is why you see 2 smaller CompileTime. Which version of pytorch/xla are you using and does the code eventually runs after compilation finishes?

**Comment by MikeynJerry** [comment]
Thanks for the prompt response, I appreciate it!

As for what you've requested, here's the HLO dump (I had to split it into 4 sequential files to meet the upload size limit):
- [hlo_tpu0_1.txt](https://github.com/pytorch/xla/files/11557695/hlo_tpu0_1.txt)
- [hlo_tpu0_2.txt](https://github.com/pytorch/xla/files/11557696/hlo_tpu0_2.txt)
- [hlo_tpu0_3.txt](https://github.com/pytorch/xla/files/11557698/hlo_tpu0_3.txt)
- [hlo_tpu0_4.txt](https://github.com/pytorch/xla/files/11557697/hlo_tpu0_4.txt)

I reran my script with these environment variables: `XLA_IR_DEBUG=True XLA_SAVE_TENSORS_FILE=ir.txt XLA_METRICS_FILE=metrics.txt XLA_SAVE_HLO_FILE=hlo.txt XLA_HLO_DEBUG=True XLA_SAVE_TENSORS_FMT=hlo`

I am using `torch==2.0.0` and `torch-xla==2.0` (the rest of my environment is [here](https://github.com/pytorch/xla/files/11557399/packages.txt)). After the code compiles three times, it does continue running / training.

Here's an updated metrics log to match the new run as well: [metrics.0.txt](https://github.com/pytorch/xla/files/11557647/metrics.0.txt)

**Comment by MikeynJerry** [comment]
@JackCaoG any ideas what might be causing this? The difference in compilation time between running with `loss.backward()` and without it is a night and day difference. With takes 30+ minutes, without takes about a couple of minutes.

Without:
```
Metric: CompileTime
  TotalSamples: 4
  Accumulator: 02m25s858ms341.533us
  ValueRate: 01s021ms120.459us / second
  Rate: 0.0281964 / second
  Percentiles: 1%=334ms868.672us; 5%=334ms868.672us; 10%=334ms868.672us; 20%=334ms868.672us; 50%=01m07s581ms329.901us; 80%=01m14s369ms873.719us; 90%=01m14s369ms873.719us; 95%=01m14s369ms873.719us; 99%=01m14s369ms873.719us
Metric: ExecuteTime
  TotalSamples: 10
  Accumulator: 788ms233.446us
  ValueRate: 006ms538.100us / second
  Rate: 0.0702596 / second
  Percentiles: 1%=001ms327.780us; 5%=001ms327.780us; 10%=006ms995.969us; 20%=041ms983.587us; 50%=079ms534.714us; 80%=151ms009.577us; 90%=166ms405.956us; 95%=166ms405.956us; 99%=166ms405.956us
```

With (and gets stuck after 5 iterations):
```
Metric: CompileTime
  TotalSamples: 4
  Accumulator: 28m01s567ms060.549us
  ValueRate: 982ms455.025us / second
  Rate: 0.00233839 / second
  Percentiles: 1%=328ms497.462us; 5%=328ms497.462us; 10%=328ms497.462us; 20%=328ms497.462us; 50%=13m11s169ms370.914us; 80%=15m46s537ms757.124us; 90%=15m46s537ms757.124us; 95%=15m46s537ms757.124us; 99%=15m46s537ms757.124us
Metric: ExecuteTime
  TotalSamples: 10
  Accumulator: 03m12s835ms715.611us
  ValueRate: 108ms003.501us / second
  Rate: 0.00563003 / second
  Percentiles: 1%=002ms897.750us; 5%=002ms897.750us; 10%=007ms146.639us; 20%=215ms744.202us; 50%=243ms281.909us; 80%=01m04s621ms067.108us; 90%=01m04s736ms264.099us; 95%=01m04s736ms264.099us; 99%=01m04s736ms264.099us
```

**Comment by JackCaoG** [comment]
without the `loss.backward` we are only compiling the forward graph, which essentially you are doing inference instead of compilation. I suspect the complexity is from the backwards pass. Let me follow up with XLA team.

**Comment by JackCaoG** [comment]
I took a quick look and the HLO dump is 70K lines in the training case, it is a lot larger than our typical graph. Is this for a single step of stable diffusion? What's the parameter size of this SD?

**Comment by MikeynJerry** [comment]
@JackCaoG 
1. Unless I've setup my script wrong, yes, I think it should be just one step of stable diffusion.

2. It's about 1B in total split between three models:
  - CLIP text encoder ~= 123M (no gradients)
  - VAE ~= 84M (no gradients)
  - UNet ~= 860M (receives gradients)

**Comment by ssusie** [comment]
Hi @MikeynJerry , for this model ~25-30 minitus of compilation is expected, later you get much faster execution. If you let the model run for few epochs and measure the time it takes to run an epoch, you can see the speedup. With TPUs you usually get slow compilation but fast execution. Also, to speedup your code, you need to reduce logging frequency (or remove it), disable progress bar update and other lines of code that do device-host communication.

Re the unlowered op, it is lowered and available in the nightly version, however using it reduces the batch size that you can fit in the device memory. Hope this helps.

**Comment by JackCaoG** [comment]
Thanks, I guess what we really want here is the persistent compilation cache so you don't need to recompiling on every call.
